sklearn/metrics/_dist_metrics.pyx.tp

```diff
@@ -2785,6 +2794,60 @@ cdef class HaversineDistance{{name_suffix}}(DistanceMetric{{name_suffix}}):
 
         return (sin_0 * sin_0 + cos(x1_0) * cos(x2_0) * sin_1 * sin_1)
 
+#------------------------------------------------------------
+# Arccos Distance
+#  D(x, y) = arccos(dot(x, y) / (|x| * |y|)) / PI
+# This is the metric version of cosine distance, i.e. the metric
+# most suitable when you wish to use cosine distance.
+
+cdef class ArccosDistance{{name_suffix}}(DistanceMetric{{name_suffix}}):
+    cdef inline {{INPUT_DTYPE_t}} cosine_similarity(
+        self,
+        {{INPUT_DTYPE_t}}* x1,
+        {{INPUT_DTYPE_t}}* x2,
+        ITYPE_t size
+    ) nogil except -1:
+        cdef {{INPUT_DTYPE_t}} d = ddot(size, x1, 1, x2, 1)
+        cdef {{INPUT_DTYPE_t}} norm1 = ddot(size, x1, 1, x1, 1)
+        cdef {{INPUT_DTYPE_t}} norm2 = ddot(size, x2, 1, x2, 1)
+        return d / sqrt(norm1 * norm2)
+
+    cdef inline {{INPUT_DTYPE_t}} rdist(
+        self,
+        {{INPUT_DTYPE_t}}* x1,
+        {{INPUT_DTYPE_t}}* x2,
+        ITYPE_t size
+    ) nogil except -1:
+        # The non-metric, but rank-preserving, cosine distance.
+        return 1 - self.cosine_similarity(x1, x2, size)
+
+    cdef inline {{INPUT_DTYPE_t}} dist(
+        self,
+        {{INPUT_DTYPE_t}}* x1,
+        {{INPUT_DTYPE_t}}* x2,
+        ITYPE_t size
+    ) nogil except -1:
+        return acos(self.cosine_similarity(x1, x2, size)) / M_PI
+
+    cdef inline {{INPUT_DTYPE_t}} _rdist_to_dist(
+        self,
+        {{INPUT_DTYPE_t}} rdist
+    ) nogil except -1:
+        return acos(1 - rdist) / M_PI
+
+    cdef inline {{INPUT_DTYPE_t}} _dist_to_rdist(
+        self,
+        {{INPUT_DTYPE_t}} dist
+    ) nogil except -1:
+        return 1 - cos(M_PI * dist)
```

Comment:
Follow-up of the previous suggestion to return `DTYPE_t` instead of `INPUT_DTYPE_t`.

```suggestion
    cdef inline DTYPE_t _rdist_to_dist(
        self,
        DTYPE_t rdist
    ) nogil except -1:
        return acos(1 - rdist) / M_PI

    cdef inline DTYPE_t _dist_to_rdist(
        self,
        DTYPE_t dist
    ) nogil except -1:
        return 1 - cos(M_PI * dist)
```